src/compiler/checker.ts

```diff
@@ -16841,15 +16841,22 @@ namespace ts {
                 return getNarrowedType(type, targetType, assumeTrue, isTypeDerivedFrom);
             }
 
-            function getNarrowedType(type: Type, candidate: Type, assumeTrue: boolean, isRelated: (source: Type, target: Type) => boolean) {
+            function getNarrowedType(type: Type, candidate: Type, assumeTrue: boolean, isRelated: (source: Type, target: Type) => boolean): Type {
                 if (!assumeTrue) {
                     return filterType(type, t => !isRelated(t, candidate));
                 }
-                // If the current type is a union type, remove all constituents that couldn't be instances of
-                // the candidate type. If one or more constituents remain, return a union of those.
                 if (type.flags & TypeFlags.Union) {
-                    const assignableType = filterType(type, t => isRelated(t, candidate));
-                    if (!(assignableType.flags & TypeFlags.Never)) {
+                    const assignableType = candidate.flags & TypeFlags.Union

```

Comment:
Why not define it recursively?:
```ts
if (candidate.flags & TypeFlags.Union) {
  return mapType(candidate, t => getNarrowedType(type, t, assumeTrue, isRelated);
}
```

Reply:
Wait, durr, `getNarrowedType` won't return never so you can't. Nvm.

Reply:
That's what I tried first, but the intersection logic at the end screws up the computation for object types. If two constituents are entirely not related, we want to drop them out of the union, not include their intersection.

Reply:
Wait, why?
```
let a: {x: number} | { y: string };
declare function isEightOrString(n: any): n is {x: 8} | {x: string}
if (isEightOrString(a)) {
  a; // Should totally be `({ x: number; } & { x: 8; }) | ({ x: number; } & { x: string; }) | ({ y: string; } & { x: 8; }) | ({ y: string; } & { x: string; })` even though `{x: string}` and `{ y: string }` are unrelated (and yes, the first one of those could supertype reduce to just `{ x: 8; }`)
}
```

Reply:
Yeah... I see that's technically correct, but we have several baselines that expect unrelated constituents to be discarded. From `partiallyDiscriminatedUnions.ts`:

```ts
class Square { kind: "square"; }
class Circle { kind: "circle"; }

type Shape = Circle | Square;
type Shapes = Shape | Array<Shape>;

declare function isShape(s : Shapes): s is Shape;

if (isShape(s)) {
    s; // Before: Shape
       // With recursive `getNarrowedType()`: Square | Circle | (Square & Shape[]) | (Circle & Shape[])
    if (s.kind === "circle") {
        s; // Before: Circle.
           // With recursive `getNarrowedType()`: Circle | (Circle & Shape[])
    }
}
```

And there's nothing technically incorrect about that, but it's definitively grosser, and we've been rolling with the current behavior for 3 years and nobody has complained.

Reply:
The above example I wrote behaves as I described _today_, though (you can put it in the playground and it yields exactly that type). And the "unrelated variants are discards" shtick is limited solely to _discriminable_ variants - types whose member types guarantee that membership to one or the other is mutually exclusive.

Reply:
> And the "unrelated variants are discarded" shtick is mostly limited solely to discriminable variants

What do you think about the Shape example then? Technically `Circle` and `Shape[]` aren't discriminable because [I can define an object that conforms to both](https://www.typescriptlang.org/play/#src=class%20Square%20%7B%20kind%3A%20%22square%22%3B%20%7D%0D%0Aclass%20Circle%20%7B%20kind%3A%20%22circle%22%3B%20%7D%0D%0A%0D%0Atype%20Shape%20%3D%20Circle%20%7C%20Square%3B%0D%0Atype%20Shapes%20%3D%20Shape%20%7C%20Array%3CShape%3E%3B%0D%0A%0D%0Aclass%20CircularShapeArray%20extends%20Array%20%7B%0D%0A%20%20%20%20kind%3A%20%22circle%22%20%3D%20%22circle%22%0D%0A%7D%0D%0A%0D%0Alet%20c%3A%20Circle%20%26%20Shape%5B%5D%20%3D%20new%20CircularShapeArray()%3B%0D%0A), but in practice it seems pretty unlikely.

There are a few other baselines that fail in the same vein, and it's even more painful when the result is an intersection between a primitive literal type and some type of object:

```ts
type S = "a" | "b";
type T = S[] | S;

function isS(t: T): t is S {
    return t === "a" || t === "b";
}

function f(foo: T) {
    if (isS(foo)) {
        return foo; // Before: S
        // After: "a" | "b" | ("a" & S[]) | ("b" & S[])
    }
}
```

I can suspend my disbelief that I might have an object that is simultaneously a `Circle` and `Shape[]`, but `"a" & S[]`? 🤨

It feels like there must be a case for using heuristics to simplify the results when a user is explicitly narrowing something. Otherwise, this function would just return the intersection no matter what, wouldn't it?

Reply:
> Otherwise, this function would just return the intersection no matter what, wouldn't it?

Intersections don't actually do supertrype reduction, so we'd still be testing to see if one side is the supertype of another and simplifying to the supertype.

But _pretty much_, yeah.

> but `"a" & S[]`

Counterexample, we have branded literals:

```ts
type Brand<TString extends string, TBrand extends object> = TString & TBrand; // Have seen this type in the wild

type PathBrand = { __path: void; };
type Path<T extends string = string> = Brand<T, PathBrand>;

declare function isPathOrEmpty(x: any): x is PathBrand | ""; // checks if string is path-y or empty

function f(foo: "/usr/home" | "/tmp") { // literals used directly instead of branded versions
    if (isPathOrEmpty(foo)) {
        return foo; // Should definitely be `("/usr/home" & PathBrand) | ("/tmp" & PathBrand)`
        // But `PathBrand` is "totally unrelated" to "/tmp" 
    }
}
```

Reply:
Sure, of course you're still right. It just _feels bad_ 😄 

I'm not sure what I would really propose here. I think most of these results appear really unintuitive, but I guess they don’t stop you from using the variable the way you would before.

And it catches an actual bug bug. [Wow.](https://www.typescriptlang.org/play/#src=interface%20A%20%7B%0D%0A%20%20%20%20foo%3A%20string%0D%0A%7D%0D%0A%0D%0Aclass%20C1%20%7B%0D%0A%20%20%20%20foo%3A%20string%3B%0D%0A%20%20%20%20c1%3A%20Date%3B%0D%0A%7D%0D%0A%0D%0Aclass%20C2%20%7B%0D%0A%20%20%20%20foo%3A%20string%3B%0D%0A%20%20%20%20c2%3A%20Date%3B%0D%0A%7D%0D%0A%0D%0Adeclare%20let%20a%3A%20A%20%7C%20C1%3B%0D%0AinitA()%3B%0D%0A%0D%0Adeclare%20function%20isC1OrC2(x%3A%20any)%3A%20x%20is%20C1%20%7C%20C2%3B%0D%0Aif%20(isC1OrC2(a))%20%7B%0D%0A%20%20%20%20a.c1.getFullYear()%3B%0D%0A%7D%0D%0A%0D%0Afunction%20initA()%20%7B%0D%0A%20%20%20%20a%20%3D%20new%20C2()%3B%0D%0A%7D)

Reply:
I've updated to this approach. You can see the handful of baselines change that look like these examples.